### ISSUE TITLE
Unable to connect

### ISSUE THREAD

**Issue by Kottakji** [kyleaa/libcluster_ec2#24]
I can't get my environment on aws elastic beanstalk to create a cluster. I am hoping somebody can help me here?

I have set up my project using a dockerfile, and deployed it to Elastic beanstalk. It's deploying fine, but it's not able to connect to other nodes. Receiving the following message:

```
20:47:23.103 [warn] [libcluster:example] unable to connect to :"my-api@172.xx.xx.xxx"
20:47:23.104 [warn] [libcluster:example] unable to connect to :"my-api@172.xx.xx.xxx"
```
There are two ec2 instances at the moment.

I added the following rules to the security groups inbound rules:
```
Custom TCP | TCP | 4369 | 0.0.0.0/0
Custom TCP | TCP | 9100 - 9155 | 0.0.0.0/0
```

Here is some other information I configured:

`env.sh.eex`
```
export RELEASE_DISTRIBUTION=name
export RELEASE_NODE=<%= @release.name %>@$(curl -s http://169.254.169.254/latest/meta-data/local-ipv4)
```

`Dockerfile`
```
FROM elixir:1.10.3-slim as builder
RUN mkdir /app
WORKDIR /app
ADD mix.* ./
RUN MIX_ENV=prod mix local.rebar
RUN MIX_ENV=prod mix local.hex --force
RUN MIX_ENV=prod mix deps.get --only prod
ADD . .
RUN MIX_ENV=prod mix release --overwrite

FROM elixir:1.10.3-slim
RUN mkdir /app
WORKDIR /app
RUN apt-get update && apt-get install -y curl
COPY --from=builder /app/_build/prod/rel/my-api .
EXPOSE 4000
CMD bin/my-api start
```

`Dockerrun.aws.json`
```
{
  "AWSEBDockerrunVersion": "1",
  "Image": {
    "Name": "xxxxxx",
    "Update": "true"
  },
  "Ports": [
    {
      "ContainerPort": "4000"
    }
  ]
}
```
`config/prod.exs`
```
config :libcluster,
       topologies: [
         example: [
           strategy: ClusterEC2.Strategy.Tags,
           config: [
             app_prefix: "my-api",
             ec2_tagname: "elasticbeanstalk:environment-name"
           ],
         ]
       ]
```

I also hardcoded the `RELEASE_COOKIE`.

**Comment by tiagodavi** [comment]
I'm having the same issue.. it looks like it's able to read data from my instances, but it couldn't connect to it

![image](https://user-images.githubusercontent.com/1058350/132587130-7c83cda5-9f7e-45d9-8c65-6979161384a9.png)
```

21:14:00.127 [warn]  [libcluster:default] unable to connect to :"app@172.31.25.170"

Interactive Elixir (1.12.2) - press Ctrl+C to exit (type h() ENTER for help)

iex(app@172.31.17.130)1> 
21:14:05.336 [debug] %ExAws.Operation.Query{action: :describe_instances, content_encoding: "identity", params: %{"Action" => "DescribeInstances", "Filter.1.Name" => "tag:web", "Filter.1.Value" => "web", "Filter.2.Name" => "instance-state-name", "Filter.2.Value" => "running", "Version" => "2016-11-15"}, parser: &ExAws.Utils.identity/2, path: "/", service: :ec2}
```

**Comment by tiagodavi** [comment]
How to allow "security groups    inter-node communication." ? I think that is what's missing.. It doesn't matter what ports I open I always end up unable to connect. It's probably something silly.

**Comment by kyleaa** [comment]
Can you SSH into the instances and see if they have connectivity between them on the ports? I think there's a lot of other VPC configuration that can impact communication between instances. The error message is a bit vague, but this doesn't seem like a libcluster issue.

https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/security-group-rules-reference.html#sg-rules-other-instances

**Comment by tiagodavi** [comment]
You mean just open inbound/outbound ports, put instances in the same subnet/vpc, same tag, same app name, same cookie is not enough? Can you remember anything that might be missing beyond what I've just said?

**Comment by tiagodavi** [comment]
Thank you anyway.

**Comment by tiagodavi** [comment]
@kyleaa Thank you very much.. now it's working with all available strategies.. The problem was always the security group. It works even with Gossip and EPMD:

![image](https://user-images.githubusercontent.com/1058350/132597422-753cbb86-90f1-4f4c-961d-297127aad20d.png)

**Comment by jorisatabix** [comment]
@tiagodavi What did you change in the end? I still haven't got it working on beanstalk.

**Comment by tiagodavi** [comment]
Hey @jorisatabix . I simply made all traffic available in my security group according to the link @kyleaa shared above. In the last column (IP Address) you can restrict it to the range of private IPs your instances are using. Something like `172.16.0.0/20` that would be the CIDR block of my case here. 

And since you are using DOCKER you need to make sure you are publishing those ports as well, otherwise even if your host is able to connect via instances they might not have access to the docker container right? I know for sure that only expose port is not enough, You need to publish them to allow external access:

https://www.whitesourcesoftware.com/free-developer-tools/blog/docker-expose-port/

**Comment by kyleaa** [comment]
That's a great callout about Docker - you would need to either run in host network mode or use port mapping to expose those. I believe that can be done in Elastic Beanstalk with the `Dockerrun.aws.json` file. In the past I had to use the Multicontainer Docker platform, I am not sure as this is currently the case as I have moved off Elastic Beanstalk in recent years.

https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/create_deploy_docker_v2config.html
